### ISSUE TITLE
Provide auto-start scripts for Windows

### ISSUE THREAD

**Issue by ngosang** [FlareSolverr/FlareSolverr#70]
We are providing a Windows executable, but it has to be started manually after each reboot.
The goal of this issue is to provide some scripts or tools to auto-start FlareSolverr.
The simplest way will be to create 2 .bat scripts and distribute them in the executable zip.
* install_autostart.bat => configure flaresolverr autostart. there are several ways in Windows. eg => https://answers.microsoft.com/en-us/windows/forum/windows_10-start-winpc/autostart-a-program-in-windows-10/940682ae-8872-47ce-964d-8b1e820d9a5a
* uninstall_autostart.bat => undo the previous changes in the system

PRs are welcome

**Comment by ilike2burnthing** [comment]
The problem with running the batch file from the startup folder is that the console is still displayed.

The two ways (outside of third-party tools) to make it run silently/hidden are:
1) the .vbs method @TheNexter used
2) a scheduled task
https://www.raymond.cc/blog/hidden-start-runs-batch-files-silently-without-flickering-console/view-all/

I use a scheduled task for my DIY Chromium updater, but I'm not sure this can be done generically (it needs the computer name, account name, and user ID, potentially also password). Happy to be corrected if wrong.
https://www.windowscentral.com/how-export-and-import-scheduled-tasks-windows-10#export_import_tasks_cmd_windows10

**Comment by ngosang** [comment]
I was thinking in a bat script which creates a scheduled task.
Maybe this works => https://store.chipkin.com/articles/writing-a-batch-script-to-add-a-windows-startup-activation-task

**Comment by ilike2burnthing** [comment]
Yea, that looks like it would work. I'll give it a try.

**Comment by ilike2burnthing** [comment]
Working - [flaresolverr.zip](https://github.com/FlareSolverr/FlareSolverr/files/5827603/flaresolverr.zip)


Names of everything can obviously be changed, and I don't know if you wanted the files in a sub-directory, just did that for the sake of tidiness. Hopefully the README makes sense as well.

**Comment by ngosang** [comment]
Thank you! I will test it and publish in the next release. Meanwhile other users can use your solution.

**Comment by ilike2burnthing** [comment]
Bump to 1.2.6 milestone

**Comment by i3laze** [comment]
I’ve used the notorious NSSM.exe to create a win32-service without any issues.
https://nssm.cc

**Comment by nydave69** [comment]
> I’ve used a well-known NSSM.exe to create a win32-service without any issues. https://nssm.cc

Last time I looked at NSSM (over a year ago) there were some major issues with it and no one seemed to be still working on it.
Hope that has changed..

**Comment by Nackophilz** [comment]
Or just Windows+R -> shell:startup -> create a shortcut inside ?

**Comment by ilike2burnthing** [comment]
Which will run with the console still displayed, hence the alternative above.

**Comment by ThePieMonster** [comment]
Was browsing for another issue and came across this. For my Windows Server 2019 environment I created a scheduled task. I exported the XML below. Feel free to include this in setup instructions for Windows users (_and maybe my name in the credits too_)  😄 
```
<?xml version="1.0" encoding="UTF-16"?>
<Task version="1.2" xmlns="http://schemas.microsoft.com/windows/2004/02/mit/task">
  <RegistrationInfo>
    <Date>2022-01-28T02:42:48.1324632</Date>
    <Author>ThePieMonster</Author>
    <URI>\FlareSolverr</URI>
  </RegistrationInfo>
  <Triggers>
    <BootTrigger>
      <Enabled>true</Enabled>
      <Delay>PT1M</Delay>
    </BootTrigger>
  </Triggers>
  <Principals>
    <Principal id="Author">
      <UserId>S-1-5-18</UserId>
      <RunLevel>LeastPrivilege</RunLevel>
    </Principal>
  </Principals>
  <Settings>
    <MultipleInstancesPolicy>IgnoreNew</MultipleInstancesPolicy>
    <DisallowStartIfOnBatteries>true</DisallowStartIfOnBatteries>
    <StopIfGoingOnBatteries>true</StopIfGoingOnBatteries>
    <AllowHardTerminate>true</AllowHardTerminate>
    <StartWhenAvailable>false</StartWhenAvailable>
    <RunOnlyIfNetworkAvailable>false</RunOnlyIfNetworkAvailable>
    <IdleSettings>
      <StopOnIdleEnd>true</StopOnIdleEnd>
      <RestartOnIdle>false</RestartOnIdle>
    </IdleSettings>
    <AllowStartOnDemand>true</AllowStartOnDemand>
    <Enabled>true</Enabled>
    <Hidden>false</Hidden>
    <RunOnlyIfIdle>false</RunOnlyIfIdle>
    <WakeToRun>false</WakeToRun>
    <ExecutionTimeLimit>PT72H</ExecutionTimeLimit>
    <Priority>7</Priority>
  </Settings>
  <Actions Context="Author">
    <Exec>
      <Command>C:\Temp\flaresolverr\flaresolverr.exe</Command>
    </Exec>
  </Actions>
</Task>
```

**Comment by beanp02** [comment]
> Working - [flaresolverr.zip](https://github.com/FlareSolverr/FlareSolverr/files/5827603/flaresolverr.zip)
> 
> Names of everything can obviously be changed, and I don't know if you wanted the files in a sub-directory, just did that for the sake of tidiness. Hopefully the README makes sense as well.

THANK YOU!!! for making it simple, took me nearly an hour to find this information surely the provider would put this info in the install instructions but anyway...

**Comment by beanp02** [comment]
Here's what i did-

A step-by-step guide to help you install Prowlarr and configure Flaresolverr using NSSM (Non-Sucking Service Manager):

Step 1: Install Prowlarr

Start by downloading the latest version of Prowlarr from the official GitHub repository: https://github.com/Prowlarr/Prowlarr/releases.
Choose the appropriate installer based on your operating system (Windows, macOS, or Linux).
Run the installer and follow the on-screen instructions to complete the installation.
Once the installation is complete, Prowlarr will launch automatically. Note down the URL and port number provided.

Step 2: Install Flaresolverr

Go to the Flaresolverr GitHub repository: https://github.com/flaresolverr/flaresolverr.
Click on the "Code" button and select "Download ZIP" to download the latest version of Flaresolverr.
Extract the downloaded ZIP file to a location of your choice (I used C:\ProgramData)

Step 3: Install NSSM

Download NSSM from the official website: https://nssm.cc/download.
Choose the appropriate installer based on your system architecture (32-bit or 64-bit).
Run the installer and follow the on-screen instructions to complete the installation.

Step 5: Configure Flaresolverr as a Windows Service using NSSM

Open a command prompt with administrative privileges.
Navigate to the directory where you installed NSSM (e.g., C:\nssm).
Execute the following command to install Flaresolverr as a service:
swift
Copy code
nssm install Flaresolverr "C:\path\to\flaresolverr.exe" -c "C:\path\to\config.json"

Replace "C:\path\to\flaresolverr.exe" with the actual path to the Flaresolverr executable and "C:\path\to\config.json" with the path to the config.json file you configured earlier.
NSSM will open a window to configure the service.
Set the "Service name" field to "Flaresolverr".
In the "Startup" tab, set the "Path" field to the Flaresolverr executable path.
In the "Arguments" field, add -c "C:\path\to\config.json" after the executable path (same as the command used in step 3).
Configure any other settings as desired.
Click the "Install service" button to create the service.
Close the NSSM window.

Step 6: Start the Flaresolverr Service

Open the Windows Services manager by pressing Windows Key + R, typing services.msc, and hitting Enter.
Look for the "Flaresolverr" service in the list.
Right-click on the service and select "Start" to initiate Flaresolverr.

Flaresolverr should now be running as a service, and you can configure Prowlarr to use it for resolving CAPTCHAs

Step 6:  Configure Flaresolverr in Prowlarr

1. Open a web browser and enter the URL and port number of your Prowlarr installation (e.g., `http://localhost:9696`).
2. Log in to Prowlarr using your credentials.
3. Click on the wrench icon in the top-right corner to access the settings.
4. In the left sidebar, select "Indexers".
5. Click on the "+ Add" button to add a new download client.
6. In the "Add Download Client" window, select "Flaresolverr" from the dropdown list of available download clients.
7. Configure the following settings for Flaresolverr:
   - Name: Flaresolverr
   - Tags: Flaresolverr
   - Host- http://localhost:8191/
8. Click the "Test" button to verify the connection to Flaresolverr. If the test is successful, you should see a green checkmark.
9. Once the connection test is successful, click the "Save" button to save the Flaresolverr configuration in Prowlarr.

Flaresolverr is now configured in Prowlarr, and it will be used for resolving CAPTCHAs during the downloading process.

**Comment by gabryk91** [comment]
I used https://github.com/winsw/winsw 
with this config:
<service>
  <id>Flaresoverr</id>
  <name>Flaresolverr</name>
  <description>Flaresolverr</description>
  <executable>c:\Flaresolverr\Flaresolverr.exe</executable>
  <logpath>c:\Flaresolverr\</logpath>
  <logmode>roll</logmode>
  <depend></depend>
  <startargument>-p</startargument>
  <startargument>c:\Flaresolverr</startargument>
  <stopexecutable>c:\Flaresolverr\Flaresolverr.exe</stopexecutable>
  <stopargument>-p</stopargument>
  <stopargument>c:\Flaresolverr</stopargument>
  <stopargument>-s</stopargument>
  <stopargument>stop</stopargument>
</service>

**Comment by iPaulis** [comment]
> config.json

Where is this json file you are referring to? how should I configure it?

**Comment by beanp02** [comment]
> > config.json
> 
> @beanp02 Where is this json file you are referring to? how should I configure it?

They are located in the files downloaded from the repo- https://github.com/FlareSolverr/FlareSolverr/releases/tag/v3.3.6 

make sure you extract the files under step 2 and record where you store them- Step 2: Install Flaresolverr

Go to the Flaresolverr GitHub repository: https://github.com/flaresolverr/flaresolverr.
Click on the "Code" button and select "Download ZIP" to download the latest version of Flaresolverr.
Extract the downloaded ZIP file to a location of your choice (I used C:\ProgramData)

**Comment by iPaulis** [comment]
Hi, thank you for your help, but I don't see any config.json in the repo.
Is it the one called [package.json](https://github.com/FlareSolverr/FlareSolverr/blob/master/package.json)?

**Comment by B-Mqn** [comment]
> I used https://github.com/winsw/winsw with this config:
> 
> ```
> <service>
>   <id>Flaresoverr</id>
>   <name>Flaresolverr</name>
>   <description>Flaresolverr</description>
>   <executable>c:\Flaresolverr\Flaresolverr.exe</executable>
>   <logpath>c:\Flaresolverr\</logpath>
>   <logmode>roll</logmode>
>   <depend></depend>
>   <startargument>-p</startargument>
>   <startargument>c:\Flaresolverr</startargument>
>   <stopexecutable>c:\Flaresolverr\Flaresolverr.exe</stopexecutable>
>   <stopargument>-p</stopargument>
>   <stopargument>c:\Flaresolverr</stopargument>
>   <stopargument>-s</stopargument>
>   <stopargument>stop</stopargument>
> </service>
> ```

worked a treat thanks @gabryk91 
hopefully it auto starts

**Comment by wmortume** [comment]
I created an auto-start windows service for FlareSolverr with C# which will also set the host env var on launch (Jackett won't work with the default one). It will work right out of the box but you can modify/append the commands to your liking. You can download the executable here: https://github.com/wmortume/FlaresolverrService/releases/tag/v1.0.0

This is how I configured the service class:

```
namespace FlaresolverrService
{
    public partial class FlaresolverrService : ServiceBase
    {
        public FlaresolverrService()
        {
            InitializeComponent();
        }

        protected override void OnStart(string[] args)
        {
            try
            {
                ProcessStartInfo processStartInfo = new ProcessStartInfo("cmd.exe", "/v /c \"set \"HOST=127.0.0.1\" && echo !HOST! && flaresolverr.exe\"") //application, command
                {
                    UseShellExecute = false,
                    WorkingDirectory = Path.Combine(Path.GetPathRoot(Environment.GetFolderPath(Environment.SpecialFolder.System)), "flaresolverr") //assumes the folder resides in the c: drive directory
                };
                Process process = Process.Start(processStartInfo);
                process.WaitForExit((int)TimeSpan.FromSeconds(15).TotalMilliseconds); //give cmd 15 secs to execute commands before closing it to prevent service from getting stuck on start status
            }
            catch (Exception ex)
            {
                throw new Exception("There was an issue with starting Flaresolverr", ex);
            }
        }

        protected override void OnStop()
        {
            foreach (var process in Process.GetProcessesByName("flaresolverr"))
            {
                process.Kill();
            }
        }
    }
}
```

The code in `ProjectInstaller.cs` allows it to auto-start without you having to open up the Services app:

```
namespace FlaresolverrService
{
    [RunInstaller(true)]
    public partial class ProjectInstaller : Installer
    {
        public ProjectInstaller()
        {
            InitializeComponent();
            serviceInstaller1.AfterInstall += ServiceInstaller_AfterInstall;
        }

        private void ServiceInstaller_AfterInstall(object sender, InstallEventArgs e) //automatically start service after installation
        {
            ServiceController serviceController = new ServiceController(serviceInstaller1.ServiceName);
            serviceController.Start();
        }
    }
}
```
I configured the `InitializeComponent()` method in `ProjectInstaller.Designer.cs` to use the local account instead of credentials and set the start mode to automatic:
```
private void InitializeComponent()
{
    this.serviceProcessInstaller1 = new System.ServiceProcess.ServiceProcessInstaller();
    this.serviceInstaller1 = new System.ServiceProcess.ServiceInstaller();
    // 
    // serviceProcessInstaller1
    // 
    this.serviceProcessInstaller1.Account = ServiceAccount.LocalSystem; //changed from requiring cred to use local account
    // 
    // serviceInstaller1
    // 
    this.serviceInstaller1.ServiceName = "FlaresolverrService";
    this.serviceInstaller1.DisplayName = "FlaresolverrService";
    this.serviceInstaller1.Description = "Flaresolverr Auto-Start Service";
    this.serviceInstaller1.StartType = ServiceStartMode.Automatic; //changed from manual to automatic
    // 
    // ProjectInstaller
    // 
    this.Installers.AddRange(new System.Configuration.Install.Installer[] {
    this.serviceProcessInstaller1,
    this.serviceInstaller1});
}
```

**Comment by orestis150** [comment]
> I created an auto-start windows service for FlareSolverr with C# which will also set the host env var on launch (Jackett won't work with the default one). It will work right out of the box but you can modify/append the commands to your liking. You can download the executable here: https://github.com/wmortume/FlaresolverrService/releases/tag/v1.0.0
> 
> This is how I configured the service class:
> 
> ```
> namespace FlaresolverrService
> {
>     public partial class FlaresolverrService : ServiceBase
>     {
>         public FlaresolverrService()
>         {
>             InitializeComponent();
>         }
> 
>         protected override void OnStart(string[] args)
>         {
>             try
>             {
>                 ProcessStartInfo processStartInfo = new ProcessStartInfo("cmd.exe", "/v /c \"set \"HOST=127.0.0.1\" && echo !HOST! && flaresolverr.exe\"") //application, command
>                 {
>                     UseShellExecute = false,
>                     WorkingDirectory = Path.Combine(Path.GetPathRoot(Environment.GetFolderPath(Environment.SpecialFolder.System)), "flaresolverr") //assumes the folder resides in the c: drive directory
>                 };
>                 Process process = Process.Start(processStartInfo);
>                 process.WaitForExit((int)TimeSpan.FromSeconds(15).TotalMilliseconds); //give cmd 15 secs to execute commands before closing it to prevent service from getting stuck on start status
>             }
>             catch (Exception ex)
>             {
>                 throw new Exception("There was an issue with starting Flaresolverr", ex);
>             }
>         }
> 
>         protected override void OnStop()
>         {
>             foreach (var process in Process.GetProcessesByName("flaresolverr"))
>             {
>                 process.Kill();
>             }
>         }
>     }
> }
> ```
> 
> The code in `ProjectInstaller.cs` allows it to auto-start without you having to open up the Services app:
> 
> ```
> namespace FlaresolverrService
> {
>     [RunInstaller(true)]
>     public partial class ProjectInstaller : Installer
>     {
>         public ProjectInstaller()
>         {
>             InitializeComponent();
>             serviceInstaller1.AfterInstall += ServiceInstaller_AfterInstall;
>         }
> 
>         private void ServiceInstaller_AfterInstall(object sender, InstallEventArgs e) //automatically start service after installation
>         {
>             ServiceController serviceController = new ServiceController(serviceInstaller1.ServiceName);
>             serviceController.Start();
>         }
>     }
> }
> ```
> 
> I configured the `InitializeComponent()` method in `ProjectInstaller.Designer.cs` to use the local account instead of credentials and set the start mode to automatic:
> 
> ```
> private void InitializeComponent()
> {
>     this.serviceProcessInstaller1 = new System.ServiceProcess.ServiceProcessInstaller();
>     this.serviceInstaller1 = new System.ServiceProcess.ServiceInstaller();
>     // 
>     // serviceProcessInstaller1
>     // 
>     this.serviceProcessInstaller1.Account = ServiceAccount.LocalSystem; //changed from requiring cred to use local account
>     // 
>     // serviceInstaller1
>     // 
>     this.serviceInstaller1.ServiceName = "FlaresolverrService";
>     this.serviceInstaller1.DisplayName = "FlaresolverrService";
>     this.serviceInstaller1.Description = "Flaresolverr Auto-Start Service";
>     this.serviceInstaller1.StartType = ServiceStartMode.Automatic; //changed from manual to automatic
>     // 
>     // ProjectInstaller
>     // 
>     this.Installers.AddRange(new System.Configuration.Install.Installer[] {
>     this.serviceProcessInstaller1,
>     this.serviceInstaller1});
> }
> ```

Goated! Now, if only the dev integrated it with the flaresolverr..

**Comment by kevindevm** [comment]
on windows with cmd and nssm:

serviceFlaresolverr.bat can be in anywhere
```
cd /d "c:\repos"

curl -L -o flaresolverr.zip ^
https://github.com/FlareSolverr/FlareSolverr/releases/latest/download/flaresolverr_windows_x64.zip

:: Crear carpeta destino y descomprimir con PowerShell
powershell -command "Expand-Archive -Path 'flaresolverr.zip' -DestinationPath '.' -Force"

"c:\repos\flaresolverr\flaresolverr.exe"
```

next in cmd as admin `cd /d "where you have nssm.exe"`
`nssm install flaresolverr "where you have this file\serviceFlaresolverr.bat"`

now start the service
`nssm start flaresolverr`
and you have it no windows appear

**Comment by alainbryden** [comment]
Here's another solution that doesn't require any 3rd party tools or elevated privileges, and runs in the background (headless) at startup.

# Step 1
In whatever folder you extracted flaresolver.exe to, create the following `launch_hidden.jse` file (or name it whatever you want):
```js
var WshShell = new ActiveXObject("WScript.Shell");
var FSO = new ActiveXObject("Scripting.FileSystemObject");
var ScriptDir = FSO.GetParentFolderName(WScript.ScriptFullName);
WshShell.CurrentDirectory = ScriptDir;
WshShell.Run('"flaresolverr.exe"', 0, false);
```

# Step 2
Create a shortcut to this file in the "Startup" folder of your choosing (either `%USERPROFILE%\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\Startup` or `C:\ProgramData\Microsoft\Windows\Start Menu\Programs\StartUp`)

---

*edit* Updated from .vbs to .jse, because Microsoft is deprecating VBScript soon, but JSE is still on life-support for now, _(Note that powershell (.ps1) scripts are not supported in the "startup" folder - hence the use of legacy scripting languages)_
